crates/sargon/src/profile/mfa/security_structures/roles/builder/roles_builder.rs

```diff
@@ -0,0 +1,857 @@
+    fn forever_invalid(reason: ForeverInvalidReason) -> Self {
+        Err(ForeverInvalid(reason))
+    }
+}
+
+pub(crate) trait FromNotYetValid {
+    fn not_yet_valid(reason: NotYetValidReason) -> Self;
+}
+impl<T> FromNotYetValid for std::result::Result<T, RoleBuilderValidation> {
+    fn not_yet_valid(reason: NotYetValidReason) -> Self {
+        Err(NotYetValid(reason))
+    }
+}
+
+pub(crate) trait FromBasicViolation {
+    fn basic_violation(reason: BasicViolation) -> Self;
+}
+impl<T> FromBasicViolation for std::result::Result<T, RoleBuilderValidation> {
+    fn basic_violation(reason: BasicViolation) -> Self {
+        Err(BasicViolation(reason))
+    }
+}
+
+impl ForeverInvalidReason {
+    pub(crate) fn threshold_list_not_supported_for_role(
+        role: RoleKind,
+    ) -> Self {
+        match role {
+            RoleKind::Recovery => {
+                Self::RecoveryRoleThresholdFactorsNotSupported
+            }
+            RoleKind::Confirmation => {
+                Self::ConfirmationRoleThresholdFactorsNotSupported
+            }
+            RoleKind::Primary => {
+                unreachable!("Primary role DOES support threshold list. This is programmer error.")
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct FactorSourceInRoleBuilderValidationStatus {
+    pub role: RoleKind,
+    pub factor_source_id: FactorSourceID,
+    pub validation: RoleBuilderMutateResult,
+}
+
+impl FactorSourceInRoleBuilderValidationStatus {
+    pub(crate) fn new(
+        role: RoleKind,
+        factor_source_id: FactorSourceID,
+        validation: RoleBuilderMutateResult,
+    ) -> Self {
+        Self {
+            role,
+            factor_source_id,
+            validation,
+        }
+    }
+}
+
+#[cfg(test)]
+impl FactorSourceInRoleBuilderValidationStatus {
+    pub(crate) fn ok(role: RoleKind, factor_source_id: FactorSourceID) -> Self {
+        Self::new(role, factor_source_id, Ok(()))
+    }
+
+    pub(crate) fn forever_invalid(
+        role: RoleKind,
+        factor_source_id: FactorSourceID,
+        reason: ForeverInvalidReason,
+    ) -> Self {
+        Self::new(
+            role,
+            factor_source_id,
+            RoleBuilderMutateResult::forever_invalid(reason),
+        )
+    }
+
+    pub(crate) fn not_yet_valid(
+        role: RoleKind,
+        factor_source_id: FactorSourceID,
+        reason: NotYetValidReason,
+    ) -> Self {
+        Self::new(
+            role,
+            factor_source_id,
+            RoleBuilderMutateResult::not_yet_valid(reason),
+        )
+    }
+}
+
+use BasicViolation::*;
+use ForeverInvalidReason::*;
+use NotYetValidReason::*;
+use RoleKind::*;
+
+pub type RoleBuilderMutateResult = Result<(), RoleBuilderValidation>;
+
+pub enum Assert<const CHECK: bool> {}
+pub trait IsTrue {}
+impl IsTrue for Assert<true> {}
+
+impl<const R: u8> RoleBuilder<R>
+where
+    Assert<{ R == ROLE_PRIMARY }>: IsTrue,
+{
```

Comment:
So this implements the below functions only if the RoleBuilder is for primary role? Seems quite convoluted that we need to define the intermediary type and trait.

Reply:
Const functionality is a bit unsupported since it is unstable feature, so one has to define those oneselves. A few lines but allows for nice `where`.